### ISSUE TITLE
smtp.SMTPClient doesn't reset timeout while sending message

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3641]
|<img alt="atsuoi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| atsuoi reported|
|-|-|
|Trac ID|trac#3641|
|Type|enhancement|
|Created|2009-02-02 08:40:08Z|
|Branch|https://github.com/twisted/twisted/tree/smtpclient-untimeout-3641|

When sending a large mail on slow connection, timeout will occur. SMTPClient should call resetTimeout() when a chunk of data was sent.

Attachments:

* [resettimeout.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e0d/e0d4724d2fbc3a418c52d28381afb58bad7d37ef/6a3748bcc9f9a93e2b2b3e378f28e0eead7fe75a.diff) (1659 bytes) - added by atsuoi on 2009-02-02 08:41:15Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3641 3641
type__enhancement enhancement
reporter__atsuoi atsuoi
priority__normal normal
milestone__ 
branch__branches_smtpclient_untimeout_3641 branches/smtpclient-untimeout-3641
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1233564008000000 1233564008000000
changetime__1233853103000000 1233853103000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="atsuoi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|atsuoi removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

The attached patch seems corrupt or invalid.  When submitting patches, it's best to check out Twisted trunk@HEAD, develop the patch against that code, then use `svn diff` to generate a patch.

For now, I'll try to apply this patch manually.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

The fix basically appears correct to me, although it's too bad that the only simple way to do this is to insert the reset into the `FileSender` hook.  It'd be nice if `FileSender` supported timeouts, perhaps.  Anyway, it's fine as-is.

The tests aren't ideal, though.  Instead, they should use `twisted.internet.task.Clock` to simulate the passage of time.  This makes them deterministic and makes them run faster.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [26162]) Branching to 'smtpclient-untimeout-3641'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [26163]) avoid timing out while sending data to the server

refs [#3641](https://github.com/twisted/twisted/issues/3641)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Oops, I meant to unassign this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

Make sure to update the copyright headers with 2009 for the files you modified.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun removed owner|
|-|-|

Updated in r26181

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @exarkun|
|-|-|

Hum, it seems it abuses the API a little bit by resetting the timeout in transform...

OTOH, I don't anything much better to propose, so please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [26189]) Merge smtpclient-untimeout-3641

Author: exarkun
Reviewer: therve
Fixes: [#3641](https://github.com/twisted/twisted/issues/3641)

Reset the idle timeout used by `SMTPClient` while sending a message body to
a server each time another chunk of bytes are written.  This prevents the
client from timing out while it's actually sending data.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
